File: backend/src/compilers/examples/bernoulli_numbers.hs
{-|
Module      : BernoulliNumbers
Description : Ada Lovelace's Bernoulli Numbers Algorithm (Note G, 1843)
Copyright   : (c) ancient-compute, 2025
License     : CC-BY-4.0
Maintainer  : ancient-compute contributors

This module implements the first computer program ever published, from Ada Lovelace's
Note G in her translation of Luigi Menabrea's memoir on the Analytical Engine.

Historical Context:
  - Published: September 1843 in Taylor's Scientific Memoirs
  - Author: Augusta Ada Byron King, Countess of Lovelace
  - Significance: First algorithm intended for machine execution
  - Original: Designed for Babbage's Analytical Engine (never built)

The Bernoulli numbers B_n are defined recursively:
    B₀ = 1
    B₁ = -1/2
    Bₙ = 0 for odd n > 1
    Bₙ = -1/(n+1) × Σ(k=0 to n-1) [C(n+1, k) × Bₖ] for even n > 1

Haskell Implementation Features:
  - Type-safe rational arithmetic using Data.Ratio
  - Pure functional implementation with pattern matching
  - Lazy evaluation for efficient computation
  - Memoization through lazy infinite list
  - Type-level guarantees preventing common errors

Ancient-Compute Integration:
  This implementation demonstrates Haskell's elegance for mathematical algorithms
  while preserving Ada's original logic. Shows how modern type systems provide
  safety guarantees that mechanical computation could not.

Compilation:
  ghc -O2 -o bernoulli bernoulli_numbers.hs

Usage:
  ./bernoulli

References:
  - Lovelace, A. (1843). "Notes by the Translator"
  - Bird, R. & Wadler, P. (1988). "Introduction to Functional Programming"
-}

module BernoulliNumbers
    ( bernoulliNumber
    , bernoulliNumberMemoized
    , bernoulliSequence
    , bernoulliAsDouble
    , printBernoulliTable
    , adaLovelaceNoteGExample
    , runTests
    ) where

import Data.Ratio (Ratio, (%), numerator, denominator)
import Data.List (genericIndex)
import Text.Printf (printf)

-- | Type alias for rational numbers (exact arithmetic)
type Rational' = Ratio Integer

-- | Calculate binomial coefficient C(n, k) = n! / (k! * (n-k)!)
--
-- Uses iterative formula to avoid factorial overflow:
-- C(n, k) = (n * (n-1) * ... * (n-k+1)) / (k * (k-1) * ... * 1)
--
-- >>> binomialCoeff 5 2
-- 10
-- >>> binomialCoeff 10 3
-- 120
binomialCoeff :: Integer -> Integer -> Integer
binomialCoeff n k
    | k < 0 || k > n = 0
    | k == 0 || k == n = 1
    | k > n - k = binomialCoeff n (n - k)  -- Optimization: C(n,k) = C(n,n-k)
    | otherwise = product [n - k + 1 .. n] `div` product [1 .. k]

-- | Calculate the nth Bernoulli number using Ada Lovelace's recursive formula.
--
-- This is the algorithm from Note G (September 1843) - the first computer program
-- ever published. Ada designed this to run on Babbage's Analytical Engine.
--
-- Mathematical Formula:
--     B₀ = 1
--     B₁ = -1/2
--     Bₙ = 0 for odd n > 1
--     Bₙ = -1/(n+1) × Σ(k=0 to n-1) [C(n+1, k) × Bₖ] for even n > 1
--
-- Complexity:
--     Time: O(n²) due to recursive summation
--     Space: O(n) for recursion depth
--
-- Examples:
--     >>> bernoulliNumber 0
--     1 % 1
--     >>> bernoulliNumber 1
--     (-1) % 2
--     >>> bernoulliNumber 2
--     1 % 6
--     >>> bernoulliNumber 8
--     (-1) % 30
--
-- Historical Note:
--     Ada's original algorithm used 25 operations on the Analytical Engine.
--     This Haskell implementation preserves the mathematical logic while using
--     pattern matching and lazy evaluation for clarity and efficiency.
bernoulliNumber :: Integer -> Rational'
bernoulliNumber n
    | n < 0     = error "Bernoulli number index must be non-negative"
    | n == 0    = 1 % 1                -- Base case: B₀ = 1
    | n == 1    = (-1) % 2             -- Base case: B₁ = -1/2
    | odd n     = 0 % 1                -- Odd Bernoulli numbers (except B₁) are zero
    | otherwise = recursiveFormula n   -- Even n > 1
  where
    -- Recursive formula: Bₙ = -1/(n+1) × Σ(k=0 to n-1) [C(n+1, k) × Bₖ]
    recursiveFormula m =
        let summation = sum [fromInteger (binomialCoeff (m + 1) k) * bernoulliNumber k
                            | k <- [0 .. m - 1]]
        in negate (summation / fromInteger (m + 1))

-- | Memoized Bernoulli numbers using lazy infinite list.
--
-- This version uses Haskell's lazy evaluation to create an infinite list
-- of Bernoulli numbers, where each element is computed only once and cached.
--
-- More efficient than the recursive version when computing multiple values.
--
-- Examples:
--     >>> bernoulliNumberMemoized 8
--     (-1) % 30
--     >>> bernoulliNumberMemoized 12
--     (-691) % 2730
bernoullisMemoized :: [Rational']
bernoullisMemoized = map bernoulliHelper [0..]
  where
    bernoulliHelper n
        | n < 0     = error "Bernoulli number index must be non-negative"
        | n == 0    = 1 % 1
        | n == 1    = (-1) % 2
        | odd n     = 0 % 1
        | otherwise = recursiveFormula n

    recursiveFormula m =
        let summation = sum [fromInteger (binomialCoeff (m + 1) k) * (bernoullisMemoized `genericIndex` k)
                            | k <- [0 .. m - 1]]
        in negate (summation / fromInteger (m + 1))

-- | Get the nth Bernoulli number from memoized list.
bernoulliNumberMemoized :: Integer -> Rational'
bernoulliNumberMemoized n
    | n < 0     = error "Bernoulli number index must be non-negative"
    | otherwise = bernoullisMemoized `genericIndex` n

-- | Generate a sequence of Bernoulli numbers from B₀ to Bₘₐₓ.
--
-- >>> bernoulliSequence 6
-- [1 % 1,(-1) % 2,1 % 6,0 % 1,(-1) % 30,0 % 1,1 % 42]
bernoulliSequence :: Integer -> [Rational']
bernoulliSequence maxN
    | maxN < 0  = error "max_n must be non-negative"
    | otherwise = take (fromIntegral maxN + 1) bernoullisMemoized

-- | Convert a Bernoulli number to Double for numerical applications.
--
-- >>> bernoulliAsDouble 2
-- 0.16666666666666666
-- >>> bernoulliAsDouble 4
-- -0.03333333333333333
bernoulliAsDouble :: Integer -> Double
bernoulliAsDouble n = fromRational (bernoulliNumber n)

-- | Print a formatted table of Bernoulli numbers.
--
-- Example output:
--     n    B_n (exact)              B_n (decimal)
--     ==================================================
--     0    1                        1.0
--     1    -1/2                     -0.5
--     2    1/6                      0.16666666666666666
printBernoulliTable :: Integer -> IO ()
printBernoulliTable maxN = do
    putStrLn "n    B_n (exact)              B_n (decimal)"
    putStrLn (replicate 70 '=')
    mapM_ printRow [0 .. maxN]
  where
    printRow n = do
        let bn = bernoulliNumberMemoized n
        when (numerator bn /= 0) $ do  -- Only print non-zero values
            let exact = show (numerator bn) ++ "/" ++ show (denominator bn)
            let exact' = if denominator bn == 1
                        then show (numerator bn)
                        else exact
            let decimal = fromRational bn :: Double
            printf "%-4d %-24s %.15f\n" n exact' decimal

    when True = const (return ())

-- | Demonstrate Ada Lovelace's Note G example.
--
-- Ada computed B₈ in her 1843 publication as the demonstration
-- of the Analytical Engine's capabilities.
adaLovelaceNoteGExample :: IO ()
adaLovelaceNoteGExample = do
    putStrLn "Ada Lovelace's Note G - Bernoulli Numbers Calculation"
    putStrLn (replicate 70 '=')
    putStrLn ""
    putStrLn "Original Context:"
    putStrLn "  Published: September 1843"
    putStrLn "  Machine: Babbage's Analytical Engine (never built)"
    putStrLn "  Variables: 9 columns (V1-V9)"
    putStrLn "  Operations: 25 steps (Load, Add, Subtract, Multiply, Divide)"
    putStrLn ""

    let n = 8
    putStrLn $ "Computing B_" ++ show n ++ " using Ada's recursive formula..."
    putStrLn ""

    let b8 = bernoulliNumberMemoized n
    putStrLn $ "Result: B_" ++ show n ++ " = " ++ formatRational b8
    putStrLn $ "Decimal: " ++ printf "%.15f" (fromRational b8 :: Double)
    putStrLn ""

    putStrLn "Verification - First 10 Bernoulli numbers:"
    printBernoulliTable 10
  where
    formatRational r
        | denominator r == 1 = show (numerator r)
        | otherwise = show (numerator r) ++ "/" ++ show (denominator r)

-- | Test suite for Bernoulli number implementation.
runTests :: IO ()
runTests = do
    putStrLn ""
    putStrLn "=== Running Bernoulli Number Tests ==="
    putStrLn ""

    let tests =
            [ ("B_0 = 1", bernoulliNumber 0 == 1 % 1)
            , ("B_1 = -1/2", bernoulliNumber 1 == (-1) % 2)
            , ("B_2 = 1/6", bernoulliNumber 2 == 1 % 6)
            , ("B_3 = 0 (odd number theorem)", bernoulliNumber 3 == 0 % 1)
            , ("B_4 = -1/30", bernoulliNumber 4 == (-1) % 30)
            , ("B_5 = 0 (odd number theorem)", bernoulliNumber 5 == 0 % 1)
            , ("B_6 = 1/42", bernoulliNumber 6 == 1 % 42)
            , ("B_8 = -1/30 (Ada's example)", bernoulliNumber 8 == (-1) % 30)
            , ("B_10 = 5/66", bernoulliNumber 10 == 5 % 66)
            , ("B_12 = -691/2730", bernoulliNumber 12 == (-691) % 2730)
            , ("B_14 = 7/6", bernoulliNumber 14 == 7 % 6)
            , ("B_20 = -174611/330", bernoulliNumber 20 == (-174611) % 330)
            , ("Memoized matches standard", bernoulliNumberMemoized 12 == bernoulliNumber 12)
            , ("Sequence length", length (bernoulliSequence 10) == 11)
            , ("All odd B_n = 0 (n=3,5,7,9)", all (\n -> bernoulliNumber n == 0 % 1) [3,5,7,9])
            ]

    let results = map runTest tests
    let passed = length (filter id results)
    let failed = length results - passed

    putStrLn ""
    putStrLn "=== Test Results ==="
    printf "Passed: %d/%d\n" passed (length results)
    printf "Failed: %d/%d\n" failed (length results)
    putStrLn ""

    if failed == 0
        then putStrLn "✓ All tests passed!"
        else putStrLn "✗ Some tests failed"
  where
    runTest (name, result) = do
        let status = if result then "✓" else "✗"
        putStrLn $ status ++ " Test: " ++ name
        result

-- | Main entry point
main :: IO ()
main = do
    -- Run Ada's historical example
    adaLovelaceNoteGExample

    -- Run test suite
    runTests

File: backend/src/compilers/examples/bernoulli_numbers.py
"""
Bernoulli Numbers Algorithm - Ada Lovelace's Note G (1843)

This module implements the first computer program ever published, from Ada Lovelace's
Note G in her translation of Luigi Menabrea's memoir on the Analytical Engine.

Historical Context:
- Published: September 1843 in Taylor's Scientific Memoirs
- Author: Augusta Ada Byron King, Countess of Lovelace
- Significance: First algorithm intended for machine execution
- Original: Designed for Babbage's Analytical Engine (never built)
- Formula: Recursive definition using binomial coefficients

The Bernoulli numbers B_n are defined recursively:
    B_0 = 1
    B_1 = -1/2
    B_n = 0 for odd n > 1
    B_n = -1/(n+1) × Σ(k=0 to n-1) [C(n+1, k) × B_k] for even n > 1

Where C(n+1, k) is the binomial coefficient "n+1 choose k".

First few values:
    B_0 = 1
    B_1 = -1/2
    B_2 = 1/6
    B_4 = -1/30
    B_6 = 1/42
    B_8 = -1/30

Ancient-Compute Integration:
This implementation is part of the historical accuracy enhancement (Phase 4),
demonstrating how ancient algorithms can be expressed in modern languages while
preserving their mathematical essence.

References:
- Lovelace, A. (1843). "Notes by the Translator" in Taylor's Scientific Memoirs, Vol. 3
- Babbage, C. (1864). "Passages from the Life of a Philosopher"
- Menabrea, L.F. (1842). "Notions sur la machine analytique de M. Charles Babbage"
"""

from fractions import Fraction
from math import comb
from typing import Union


def bernoulli_number(n: int) -> Fraction:
    """
    Calculate the nth Bernoulli number using Ada Lovelace's recursive formula.

    This is the algorithm from Note G (September 1843) - the first computer program
    ever published. Ada designed this to run on Babbage's Analytical Engine.

    Args:
        n: The index of the Bernoulli number to calculate (n >= 0)

    Returns:
        The nth Bernoulli number as a Fraction (exact rational arithmetic)

    Raises:
        ValueError: If n < 0

    Mathematical Formula:
        B_0 = 1
        B_1 = -1/2
        B_n = 0 for odd n > 1
        B_n = -1/(n+1) × Σ(k=0 to n-1) [C(n+1, k) × B_k] for even n > 1

    Complexity:
        Time: O(n²) due to recursive summation
        Space: O(n) for recursion depth

    Examples:
        >>> bernoulli_number(0)
        Fraction(1, 1)

        >>> bernoulli_number(1)
        Fraction(-1, 2)

        >>> bernoulli_number(2)
        Fraction(1, 6)

        >>> bernoulli_number(4)
        Fraction(-1, 30)

        >>> bernoulli_number(6)
        Fraction(1, 42)

    Historical Note:
        Ada's original algorithm used 25 operations on the Analytical Engine:
        - 9 columns (V1-V9) for variables
        - Operations: Load, Add, Subtract, Multiply, Divide
        - Result stored in V11

        Modern implementation uses memoization and Python's arbitrary precision
        arithmetic, but follows the same mathematical logic.
    """
    if n < 0:
        raise ValueError(f"Bernoulli number index must be non-negative, got {n}")

    # Base case: B_0 = 1
    if n == 0:
        return Fraction(1)

    # Base case: B_1 = -1/2
    if n == 1:
        return Fraction(-1, 2)

    # All odd Bernoulli numbers (except B_1) are zero
    if n > 1 and n % 2 == 1:
        return Fraction(0)

    # Recursive formula: B_n = -1/(n+1) × Σ(k=0 to n-1) [C(n+1, k) × B_k]
    summation = Fraction(0)
    for k in range(n):
        binomial_coeff = comb(n + 1, k)  # C(n+1, k)
        B_k = bernoulli_number(k)         # Recursive call
        summation += binomial_coeff * B_k

    B_n = -summation / (n + 1)
    return B_n


def bernoulli_number_memoized(n: int, memo: dict = None) -> Fraction:
    """
    Calculate the nth Bernoulli number with memoization for efficiency.

    This optimized version caches previously computed values to avoid redundant
    recursive calculations. More efficient for computing multiple Bernoulli numbers.

    Args:
        n: The index of the Bernoulli number to calculate (n >= 0)
        memo: Optional memoization dictionary (used internally for recursion)

    Returns:
        The nth Bernoulli number as a Fraction

    Complexity:
        Time: O(n²) with memoization (vs O(2^n) without)
        Space: O(n) for memoization cache

    Examples:
        >>> bernoulli_number_memoized(8)
        Fraction(-1, 30)

        >>> bernoulli_number_memoized(10)
        Fraction(5, 66)
    """
    if memo is None:
        memo = {}

    if n in memo:
        return memo[n]

    if n < 0:
        raise ValueError(f"Bernoulli number index must be non-negative, got {n}")

    if n == 0:
        result = Fraction(1)
    elif n == 1:
        result = Fraction(-1, 2)
    elif n > 1 and n % 2 == 1:
        result = Fraction(0)
    else:
        summation = Fraction(0)
        for k in range(n):
            binomial_coeff = comb(n + 1, k)
            B_k = bernoulli_number_memoized(k, memo)
            summation += binomial_coeff * B_k
        result = -summation / (n + 1)

    memo[n] = result
    return result


def bernoulli_sequence(max_n: int) -> list[Fraction]:
    """
    Generate a sequence of Bernoulli numbers from B_0 to B_max_n.

    Args:
        max_n: The highest index to calculate (inclusive)

    Returns:
        List of Bernoulli numbers [B_0, B_1, ..., B_max_n]

    Examples:
        >>> bernoulli_sequence(6)
        [Fraction(1, 1), Fraction(-1, 2), Fraction(1, 6), Fraction(0, 1),
         Fraction(-1, 30), Fraction(0, 1), Fraction(1, 42)]
    """
    if max_n < 0:
        raise ValueError(f"max_n must be non-negative, got {max_n}")

    memo = {}
    return [bernoulli_number_memoized(i, memo) for i in range(max_n + 1)]


def bernoulli_as_float(n: int, precision: int = 10) -> float:
    """
    Calculate the nth Bernoulli number and return as float.

    Useful for numerical applications where exact rational arithmetic is not needed.

    Args:
        n: The index of the Bernoulli number
        precision: Number of decimal places (for display only)

    Returns:
        Bernoulli number as floating point

    Examples:
        >>> round(bernoulli_as_float(2), 6)
        0.166667

        >>> round(bernoulli_as_float(4), 6)
        -0.033333
    """
    b_n = bernoulli_number(n)
    return float(b_n)


def print_bernoulli_table(max_n: int = 10) -> None:
    """
    Print a formatted table of Bernoulli numbers.

    Args:
        max_n: Maximum index to display

    Example Output:
        n    B_n (exact)              B_n (decimal)
        ================================================
        0    1                        1.0
        1    -1/2                     -0.5
        2    1/6                      0.16666666666666666
        4    -1/30                    -0.03333333333333333
        6    1/42                     0.023809523809523808
        8    -1/30                    -0.03333333333333333
        10   5/66                     0.07575757575757576
    """
    print("n    B_n (exact)              B_n (decimal)")
    print("=" * 70)

    for i in range(max_n + 1):
        b_n = bernoulli_number(i)
        if b_n != 0:  # Only print non-zero values
            exact = str(b_n)
            decimal = float(b_n)
            print(f"{i:<4} {exact:<24} {decimal}")


# Historical demonstration: Ada's original example from Note G
def ada_lovelace_note_g_example() -> None:
    """
    Demonstrate the calculation as described in Ada Lovelace's Note G.

    Ada's original table shows the operations needed to compute B_8 on the
    Analytical Engine using 25 operations across 9 variable columns.

    This function replicates the logic (though not the exact operational sequence)
    to compute the same result.
    """
    print("Ada Lovelace's Note G - Bernoulli Numbers Calculation")
    print("=" * 70)
    print("\nOriginal Context:")
    print("  Published: September 1843")
    print("  Machine: Babbage's Analytical Engine (never built)")
    print("  Variables: 9 columns (V1-V9)")
    print("  Operations: 25 steps (Load, Add, Subtract, Multiply, Divide)")
    print()

    # Compute B_8 using Ada's formula
    n = 8
    print(f"Computing B_{n} using Ada's recursive formula...")
    print()

    b_8 = bernoulli_number(n)

    print(f"Result: B_{n} = {b_8}")
    print(f"Decimal: {float(b_8):.15f}")
    print()

    print("Verification - First 10 Bernoulli numbers:")
    print_bernoulli_table(10)


if __name__ == "__main__":
    # Run Ada's historical example
    ada_lovelace_note_g_example()
